web/src/composebox_typeahead.js

```diff
@@ -151,38 +152,70 @@ export function should_enter_send(e) {
+    let to_append = "";
+    // if previous line was bulleted automatically add a bullet to the new line
+    if (list_util.is_bulleted(previous_line)) {
+        // if previous line had only bullet, remove it and stay on the same line
+        if (list_util.strip_bullet(previous_line) === "") {
+            // the below line replaces the last 2 characters in the textarea before
+            // the cursor - the bullet syntax - with an empty string
+            $textarea.range($textarea.caret() - 2, $textarea.caret()).range("");
+            compose_ui.autosize_textarea($textarea);
+            e.preventDefault();
+            return;
+        }
+        // use same bullet syntax as the previous line
+        to_append = previous_line.slice(0, 2);
+    } else if (list_util.is_numbered(previous_line)) {
+        // if previous line was numbered continue numbering with the new line
+        const previous_number_string = previous_line.slice(0, previous_line.indexOf("."));
+        // if previous line had only numbering, remove it and stay on the same line
+        if (list_util.strip_numbering(previous_line) === "") {
+            // the below line replaces the last few characters in the textarea before
+            // the cursor - the numbering syntax - with an empty string
+            $textarea
+                .range($textarea.caret() - previous_number_string.length - 2, $textarea.caret())
+                .range("");
+            compose_ui.autosize_textarea($textarea);
+            e.preventDefault();
+            return;
+        }
+        const previous_number = Number.parseInt(previous_number_string, 10);
+        to_append = previous_number + 1 + ". ";
+    }
+    // if previous line was neither numbered nor bulleted, only add
+    // a new line to emulate default behaviour (to_append is blank)
+    // else we add the bulleting / numbering syntax to the new line
+    $textarea.caret("\n" + to_append);
+    compose_ui.autosize_textarea($textarea);
+    e.preventDefault();
+}
+
 export function handle_enter($textarea, e) {
     // Used only if Enter doesn't send.
 
-    // Since this Enter doesn't send, we just want to do
-    // the browser's default behavior for the "Enter" key.
-    // Letting the browser handle it works great if the
-    // key actually pressed was Enter or Shift-Enter.
-
-    // But the default browser behavior for Ctrl/Alt/Meta
-    // + Enter is to do nothing, so we need to emulate
-    // the browser behavior for "Enter" in those cases.
-    //
-    // We do this using caret and range from jquery-caret.
-    const has_non_shift_modifier_key = e.ctrlKey || e.metaKey || e.altKey;
-    if (!has_non_shift_modifier_key) {
-        // Use the native browser behavior.
-        return;
-    }
-
     // To properly emulate browser "Enter", if the
     // user had selected something in the textarea,
     // we need those characters to be cleared.
     const range = $textarea.range();
```

Comment:
Should this be modernized to use the `getSelection` API too?

Reply:
After some tinkering I realised that using the `getSelection` API would be rather overkill here, and simply used textarea's properties `selectionStart` and `selectionEnd` which do the job well enough.